app/models/user.rb

```diff
@@ -0,0 +1,11 @@
+class User < ApplicationRecord
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
+
+  before_save{self.email = email.downcase}
+  validates :name,  presence: true, length: {maximum: 50}
+  validates :email, presence: true, length: {maximum: 255},
+            format: {with: VALID_EMAIL_REGEX},
+            uniqueness: {case_sensitive: false}
+  has_secure_password
```

Comment:
hàm này dùng để làm gì

Reply:
dạ để validates email của người dùng
presence : trường email là trường bắt buộc phải có.
lenght: giới hạn độ dài email không quá 255 ký tự vì email: kiểu string chỉ có 256 ký tự
format là check định dạng của email . với tham số with là một biểu thức chính quy để check địa chỉ email có đúng định dạng abc@xyz.com
uniqueness là xác thực email là duy nhất trong trường hợp email được in hoa . truyền tham số case_sensitive = false để hàm uniquensess check nếu như email hoa với email cho dù có in hoa thì nó là 1

Reply:
`has_secure_password` đang hỏi method này mà gái

Reply:
dạ has_secure_password là 1 helper của rails dùng để hash password của tài khoản . 
khi sử dụng has_secure_password đồng thời tạo 1 biến password_digest = brycen(password)
, sinh ra thêm 1 thuộc tính ảo là password_confirmation để config lại mật khẩu khi người dùng đăng ký và cái cuối cùng là cung cấp method authenticate để thuận tiện cho người sử dụng check thông tin người dùng . authencate với tham số truyền vào là chuỗi password kiểu string sau đó được mã hóa rồi so sánh với  password_digest  nếu đúng thì trả về true